spacy/pipeline/pipes.pyx

```diff
@@ -1150,49 +1152,40 @@ class EntityLinker(Pipe):
             ents_by_offset = dict()
             for ent in doc.ents:
                 ents_by_offset[str(ent.start_char) + "_" + str(ent.end_char)] = ent
-            for entity in gold.links:
-                start, end, gold_kb = entity
+            for entity, kb_dict in gold.links.items():
+                start, end = entity
                 mention = doc.text[start:end]
+                for kb_id, value in kb_dict.items():
+                    entity_encoding = self.kb.get_vector(kb_id)
+                    prior_prob = self.kb.get_prior_prob(kb_id, mention)
+
+                    gold_ent = ents_by_offset[str(ent.start_char) + "_" + str(ent.end_char)]
+                    assert gold_ent is not None
+                    type_vector = [0 for i in range(len(type_to_int))]
+                    if len(type_to_int) > 0:
+                        type_vector[type_to_int[gold_ent.label_]] = 1
 
-                gold_ent = ents_by_offset[str(ent.start_char) + "_" + str(ent.end_char)]
-                assert gold_ent is not None
-                type_vector = [0 for i in range(len(type_to_int))]
-                if len(type_to_int) > 0:
-                    type_vector[type_to_int[gold_ent.label_]] = 1
-
-                candidates = self.kb.get_candidates(mention)
-                random.shuffle(candidates)
-                nr_neg = 0
-                for c in candidates:
-                    kb_id = c.entity_
-                    entity_encoding = c.entity_vector
+                    # store data
                     entity_encodings.append(entity_encoding)
                     context_docs.append(doc)
                     type_vectors.append(type_vector)
 
                     if self.cfg.get("prior_weight", 1) > 0:
-                        priors.append([c.prior_prob])
+                        priors.append([prior_prob])
                     else:
                         priors.append([0])
 
-                    if kb_id == gold_kb:
-                        cats.append([1])
-                    else:
-                        nr_neg += 1
-                        cats.append([0])
-
         if len(entity_encodings) > 0:
-            assert len(priors) == len(entity_encodings) == len(context_docs) == len(cats) == len(type_vectors)
+            assert len(priors) == len(entity_encodings) == len(context_docs) == len(type_vectors)
```

Comment:
Todo: replace `assert` with explicit error.

Reply:
yea? In this case they should really be the same if the code is well-written, it doesn't depend on user input. So I thought `assert` would be more appropriate?

Reply:
Okay, I guess if we're 100% sure that this should never happen and it's really just an invariant, then yes. And if there's nothing the user can possibly put in that would trigger this, and not really a way to introduce a bug ourselves in some other piece of code that would have an effect. (There are a few places in the code base where we raise errors like "XY went wrong. This is likely a bug in spaCy.")

So if you think there's not really a risk here, we can leave the assert. Maybe just add a comment then so it's clear what it means if the assertion failed in this case.

Reply:
I like the idea of specifically mentioning this might be a bug in spaCy - refactoring with custom error messages.